adapters/from_sql.py

```diff
@@ -0,0 +1,187 @@
+    if most_recent_task:
+      # allows for continuation even if tasks have not begun/been generated yet
+      release.last_modified = to_timestamp(most_recent_task.last_modified)
+      release.last_active_task = most_recent_task.task_name
+    else:
+      release.last_modified = to_timestamp(item[2])
+      release.last_active_task = ''
+    release_objects[release.release_id] = release
+
+  return release_objects
+
+
+def read_tasks(task_data):
+  """Reads the tasks, transforms them into objects.
+
+  Args:
+    task_data: a tuple of tuples with task info
+
+  Returns:
+    task_objects
+  """
+  task_objects = []
+  for item in task_data:
+    task = Task()
+
+    task.task_name = item[0]
+    task.add_dependency = None  # TODO(dommarques): figure out how to get this - dag info?
+    task.started = to_timestamp(item[3])
+    task.status = state_from_string(item[6])
+    task.last_modified = to_timestamp(item[4])
+    task.log_url = 'https://youtu.be/dQw4w9WgXcQ'  # TODO(dommarques): figure out how to get the log in here
+    task.error = item[6]
+    task_objects.append(task)
+  return task_objects
+
+
+def get_task_info(execution_date, airflow_db):
+  """Gets task-related info to fill in missing release object info.
+
+  Args:
+    execution_date: a python datetime.datetime
+    airflow_db: the database connection object
+
+  Returns:
+    most_recent_task: the most recent active task
+    state: the state of the release based on the task states
+  """
+  task_query = to_sql_tasks(execution_date)
+  task_data = airflow_db.query(task_query)
+  task_objects = read_tasks(task_data)
+  state = 0
+  task_ids = []
+  if task_objects:
+    most_recent_task = task_objects[len(task_objects) - 1]
+  else:
+    most_recent_task = None
+  for task in task_objects:
+    if task.status > state:
+      state = task.status
+    task_ids.append(task.task_name)
+  return task_ids, most_recent_task, state
+
+
+def read_xcom_vars(xcom_data):
+  """Reads the xcom data to get the dict of vars and the green build SHA."""
+  # occasionally, the xcom data doesn't exist yet. This is a workaround
+  try:
+    xcom_dict = json.loads(xcom_data[0][2])
+    green_sha = xcom_data[1][2]
+    return xcom_dict, green_sha
+  except IndexError:
+    return None, None
+
+
+def get_xcom(execution_date, dag_id, airflow_db):
+  """Gets relevant xcom data from SQL.
+
+    Args:
+      execution_date: int or float
+      dag_id: str
+      airflow_db: airflow database connection object
+
+    Returns:
+      xcom_dict: the dictionary of xcom information
+      green_sha: the green build SHA
+  """
+  xcom_query = to_sql_xcom(dag_id, execution_date)
+  xcom_data = airflow_db.query(xcom_query)
+  xcom_dict, green_sha = read_xcom_vars(xcom_data)
+  return xcom_dict, green_sha
+
+
+def state_from_string(state):
+  """Transforms the state from a string format to an enumerated version.
+
+  Args:
+    state: string
+
+  Returns:
+    int
+  """
+  if state == 'none':
```

Comment:
same comment here, use the constant representing the number. You might want to define a global map for this.
```python
STATE_FROM_STRING = {
"running": State.RUNNING,
...
}
```
and then use it like:

STATE_FROM_STRING.get(state, STATE.UNKNOWN)